src/librustc_mir/interpret/validity.rs

```diff
@@ -118,6 +108,24 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M> {
                 bits
             },
             Scalar::Ptr(_) => {
+                match ty.sty {
+                    ty::Bool |
+                    ty::Char |
+                    ty::Float(_) |
+                    ty::Int(_) |
+                    ty::Uint(_) => {
+                        return validation_failure!(
+                                "a pointer",
+                                path,
+                                format!("the type {}", ty.sty)
+                            );
+                    }
+                    ty::RawPtr(_) |
+                    ty::Ref(_, _, _) |
+                    ty::FnPtr(_) => {}
+                    _ => { unreachable!(); }
```

Comment:
Wouldn't it be much easier to handle this in `validate_scalar`? There's already a type match there to special-case `Char`; that could be used to do an *extra* check for the integer types.

Reply:
I can move it into the main body if you like

Reply:
@RalfJung this is validate_scalar. The char match is when we know it's no pointer, so we can't do it there

Reply:
Oh. Not enough diff context. The never mind me.^^